packages/engine/src/datastore/table/pool/mod.rs

```diff
@@ -20,44 +20,110 @@ trait Pool<B> {
     fn get_batches_mut(&mut self) -> &mut Vec<Arc<RwLock<B>>>;
 }
 
-/// A pool is an ordered collection of batches, where each group of agents within a simulation run
-/// is associated to a batch in the pool.
+/// A pool is an ordered collection of batches.
+///
+/// Each group of agents within a simulation run is associated to a [`Batch`] in the pool. Each
+/// [`Batch`] in a pool can either be borrowed as shared ([`read_proxies()`] and
+/// [`partial_read_proxies()`] returning [`PoolReadProxy`]) or mutable reference
+/// ([`write_proxies()`] and [`partial_write_proxies()`] returning [`PoolReadProxy`]).
+///
+/// [`read_proxies()`]: Self::read_proxies
+/// [`partial_read_proxies()`]: Self::partial_read_proxies
+/// [`write_proxies()`]: Self::write_proxies
+/// [`partial_write_proxies()`]: Self::partial_write_proxies
 pub trait BatchPool<B: Batch>: Send + Sync {
+    /// Creates a new pool from [`Batches`].
+    ///
+    /// Because of the way `BatchPools` are organized it's required that the [`Batch`]es are
+    /// stored inside an [`RwLock`] behind an [`Arc`]. This is subject to change.
     fn new(batches: Vec<Arc<RwLock<B>>>) -> Self;
 
+    /// Creates a new empty pool.
     fn empty() -> Self
     where
         Self: Sized,
     {
         Self::new(Vec::new())
     }
 
+    /// Returns the number of [`Batch`]es inside this pool.
     fn len(&self) -> usize;
 
+    /// Returns true if there are no [`Batch`]es in this pool.
     fn is_empty(&self) -> bool {
         self.len() == 0
     }
 
+    /// Adds a [`Batch`] at the end of this pool.
+    ///
+    /// Note, that unlike in [`new()`](Self::new) this does not require the [`Batch`] to be wrapped
+    /// inside of [`Arc`]`<RwLock<B>>`.
     fn push(&mut self, batch: B);
 
+    /// Removes at position `index` within the pool, shifting all elements after it to the left.
+    ///
+    /// Returns the batch id of the removed [`Batch`].
+    ///
+    /// # Panics
+    ///
+    /// - If `index` is out of bounds, or
+    /// - if the `Batch` is currently borrowed as a [`BatchReadProxy`] or [`BatchWriteProxy`].
     fn remove(&mut self, index: usize) -> String;
 
+    /// Removes a [`Batch`] from the pool and returns its id.
+    ///
+    /// The removed [`Batch`] is replaced by the last [`Batch`] of the pool. This does not preserve
+    /// ordering, but is `O(1)`. If you need to preserve the element order, use
+    /// [`remove()`](Self::remove) instead.
+    ///
+    /// # Panics
+    ///
+    /// - If `index` is out of bounds, or
+    /// - if the `Batch` is currently borrowed as a [`BatchReadProxy`] or [`BatchWriteProxy`].
     fn swap_remove(&mut self, index: usize) -> String;
 
     /// Creates a [`PoolReadProxy`] for _all_ batches within the pool.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`ProxySharedLock`] if any of the [`Batch`]es is currently borrowed in a
```

Comment:
The naming of the Error is pretty confusing.
> Returns a [`ProxySharedLock`]

Makes it seem like if there's an Error it returns a Proxy with a SharedLock